text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,492 @@
+name as the parameter.
+
+## `struct`s and lifetimes
+
+Sometimes you need to build data types that contain borrowed data. Since those
+types can then be used in many contexts, you can't say in advance what the
+lifetime of those borrows will be. Instead, you must take it as a parameter:
+
+```rust
+struct VecIter<'vec, T> {
+    vec: &'vec Vec<T>,
+    index: usize,
+}
+```
+
+Here we're defining a type for iterating over a vector, without requiring
+*ownership* of that vector. To do so, we store a *borrow* of the vector. But
+because our new `VecIter` struct contains borrowed data, it needs to surface
+that fact, and the lifetime connected with it. It does so by taking an explicit
+`'vec` parameter for the relevant lifetime, and using it within.
+
+When using this struct, you can apply explicitly-named lifetimes as usual:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&'vec self) -> VecIter<'vec, T> { ... }
+}
+```
+
+However, in cases like this example, we would normally be able to leave off the
+lifetime with `&`, since there's only one source of data we could be borrowing
+from. We can do something similar with structs:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&self) -> VecIter<_, T> { ... }
+}
+```
+
+The `_` marker makes clear to the reader that *borrowing is happening*, which
+might not otherwise be clear.
+
+## `impl` blocks and lifetimes
+
+When writing an `impl` block for a structure that takes a lifetime parameter,
+you can give that parameter a name, but it must begin with an uppercase letter:
+
+```rust
+impl<T> VecIter<'Vec, T> { ... }
+```
+
+The reason for this distinction is so that there's no potential for confusion
+with lifetime names introduced by methods within the `impl` block. In other
+words, it makes clear whether the method is *introducing* a new lifetime
+(lowercase), or *referring* to the one in the `impl` header (uppercase):
+
+```rust
+impl<T> VecIter<'Vec, T> {
+    fn foo(&self) -> &'Vec T { ... }
+    fn bar(&self, arg: &'arg Bar) -> &'arg Bar { ... }
+
+    // these two are the same:
+    fn baz(&self) -> &T { ... }
+    fn baz(&'self self) -> &'self T { ... }
+}
+```
+
+If the type's lifetime is not relevant, you can leave it off using `_`:
+
+```rust
+impl<T> VecIter<_, T> { ... }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Lifetimes in `impl` headers
+
+When writing an `impl` header, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `impl<'a>`). It is also deprecated to
+use any lifetime variables that do not begin with a capital letter (which is
+needed to clearly distinguish lifetimes that are bound in the `impl` header from
+those bound in signatures).
+
+Instead the `impl` header can mention lifetimes without adding them as
+generics. **These lifetimes *always* become parameters of the `impl`**--after
+all, **no lifetime variables can possibly be in scope already**. In other words,
```

Comment:
I worry that building on the assumption that "no lifetime variables can possibly be in scope already" is going to rule out possible future language changes that would be a bigger net win.

I'm thinking in particular of lifetimes bound at the module level (which maybe would behave like functors in ML?) ... this is something that I know nmatsakis and perhaps also nrc have been tossing around ideas here ... which could be very useful for the ergonomics of large modules with heavy repetition of one or more lifetimes (the compiler is our go-to example of this)